src/io.jl

```diff
@@ -1,39 +1,81 @@
 # TODO: maybe more fancy file format and correctness checking should be done
-
+struct HGF_FORMAT end
+struct JSON_FORMAT end
 
 """
-    hg_save(io::IO, h::Hypergraph)
+    hg_save(io::IO, h::Hypergraph, format::Type{HGF_FORMAT})
 
-Saves a hypergraph `h` to an output stream `io`.
+Saves a hypergraph `h` to an output stream `io` in `hgf` format.
 
 """
-function hg_save(io::IO, h::Hypergraph)
+function hg_save(io::IO, h::Hypergraph, format::Type{HGF_FORMAT})
     println(io, length(h.v2he), " ", length(h.he2v))
     for he in h.he2v
         skeys = sort(collect(keys(he)))
         println(io, join(["$k=$(he[k])" for k in skeys], ' '))
     end
 end
 
+
+"""
+    hg_save(io::IO, h::Hypergraph, format::Type{JSON_FORMAT})
+
+Saves a hypergraph `h` to an output stream `io` in `json` format.
+The `json` in output contains the following information (keys):
+
+* `n` : number of vertices
+* `k` : number of hyperedges
+* `m` : a matrix representation of `h` where rows are vertices and columns are hyperedges
+* `v2he` : mapping vertices to hyperedges
+* `v_meta` : vertices metadata
+* `he_meta` : hyperedges metadata
+
+"""
+function hg_save(io::IO, h::Hypergraph, format::Type{JSON_FORMAT})
+    json_hg = Dict{Symbol, Any}()
+
+    json_hg[:n] = nhv(h)
+    json_hg[:k] = nhe(h)
+
+    json_hg[:m] = JSON3.write(Matrix(h))
+    json_hg[:v2he] = JSON3.write(h. v2he)
+
+    json_hg[:v_meta] = JSON3.write(h.v_meta)
+    json_hg[:he_meta] = JSON3.write(h.he_meta)
+
+    JSON3.write(io, json_hg)
+end
+
+
 """
-    hg_save(fname::AbstractString, h::Hypergraph)
+    hg_save(
+        fname::AbstractString, h::Hypergraph;
+        format::Union{Type{HGF_FORMAT}, Type{JSON_FORMAT_DICT}, Type{JSON_FORMAT_MATRIX}}
+    )
 
-Saves a hypergraph `h` to a file `fname`.
+Saves a hypergraph `h` to a file `fname` in the specified `format`.
 
 """
-hg_save(fname::AbstractString, h::Hypergraph) =
-    open(io -> hg_save(io, h), fname, "w")
+hg_save(
+    fname::AbstractString, h::Hypergraph;
+    format::Union{Type{HGF_FORMAT}, Type{JSON_FORMAT}}=HGF_FORMAT) =
+    open(io -> hg_save(io, h, format), fname, "w")
+
 
 """
-    hg_load(fname::AbstractString, T::Type{<:Real})
+    hg_load(
+        fname::AbstractString, T::Type{<:Real}, format::Type{HGF_FORMAT};
+        V=Nothing, E=Nothing
+    )
 
-Loads a hypergraph from a stream `io`. The second argument
-`T` represents type of data in the hypegraph.
+Loads a hypergraph from a stream `io` from `hgf` format. The second argument
+`T` represents type of data in the hypegraph. `V` and `E` represent the type
+of vertex metadata and hyperedge metadata, respectively.
 
-Skips an initial comment.
+Skips a single initial comment.
 
 """
-function hg_load(io::IO, T::Type{<:Real})
+function hg_load(io::IO, T::Type{<:Real}, format::Type{HGF_FORMAT}; V=Nothing, E=Nothing)
```

Comment:
- provide support for the internal storage model `D` (new functionality of the library).
- the default value of `T` is `Bool`
Hence the function should look like this
```
 function hg_load(io, T::Type{U}=Bool, D::Type{<:AbstractDict{Int, U}}=Dict{Int,U} , format::Type{HGF_FORMAT}; V=Nothing, E=Nothing) where U <: Real
```